mozetl/search/aggregates.py

```diff
@@ -171,28 +185,96 @@ def get_search_addon_version(active_addons):
     )
 
 
+def get_ad_click_count(ad_click_count):
+    if ad_click_count:
+        return [
+            {"engine": e, "source": "ad-click:", "count": c}
+            for e, c in ad_click_count.items()
+        ]
+    return []
+
+
+def get_search_with_ads_count(search_with_ads):
+    if search_with_ads:
+        return [
+            {"engine": e, "source": "search-with-ads:", "count": c}
+            for e, c in search_with_ads.items()
+        ]
+    return []
+
+
 def explode_search_counts(main_summary):
-    exploded_col_name = "single_search_count"
-    search_fields = [
-        exploded_col_name + "." + field for field in ["engine", "source", "count"]
-    ]
-
-    exploded_search_counts = (
-        main_summary.withColumn(exploded_col_name, explode(col("search_counts")))
-        .select(["*"] + search_fields)
-        .filter("single_search_count.count < %s" % MAX_CLIENT_SEARCH_COUNT)
-        .drop(exploded_col_name)
-        .drop("search_counts")
+    def _get_search_fields(exploded_col_name):
+        return [
+            exploded_col_name + "." + field for field in ["engine", "source", "count"]
+        ]
+
+    base_search_counts = main_summary.withColumn(
+        "single_search_count", explode(col("search_counts"))
+    ).filter("single_search_count.count < %s" % MAX_CLIENT_SEARCH_COUNT)
+
+    def _drop_source_columns(base):
+        for source_col in [
+            "search_counts",
+            "scalar_parent_browser_search_ad_clicks",
+            "scalar_parent_browser_search_with_ads",
+        ]:
+            base = base.drop(source_col)
+        return base
+
+    def _get_ad_counts(scalar_name, col_name, udf_function):
```

Comment:
It feels like `_get_ad_counts` is doing something very similar to the work at lines 245. Is this generalizing the explode we do on `SEARCH_COUNTS` to apply to the ad-click probes? If so, can we refactor `explode_search_counts` so that the three explodes for `SEARCH_COUNTS`, ad-clicks, and searches-with-ads are all handled consistently?

Reply:
Tried to do some refactoring into a common function. The code is rather nested and convoluted for my liking, but what can you do. :/